### PR TITLE
[stereoscopicmanager] Ensure we don't have a stale value of videoplayer stereoscopic mode

### PR DESCRIPTION
**Pull request by popcornmix**
Switching to 3D mode doesn't always work. There is a race condition when
CStereoscopicsManager::OnPlaybackStarted is called whether the stereo mode
read from infoManager is valid or not.

Ensuring the stereo_mode in VideoPlayer is valid at start of playback and
that infoManager is up to date avoids this issue.

### CODE REVIEW

**Comment by popcornmix** [comment]
Ping @FernetMenta @da-anda 
Report of fixed problem: http://forum.kodi.tv/showthread.php?tid=250817&pid=2204876#pid2204876

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 1]
```diff
@@ -3562,6 +3562,20 @@ bool CVideoPlayer::OpenVideoStream(CDVDStreamInfo& hint, bool reset)
   if (hint.stereo_mode.empty())
     hint.stereo_mode = CStereoscopicsManager::GetInstance().DetectStereoModeByString(m_item.GetPath());
 
+  {
+    SelectionStream& s = m_SelectionStreams.Get(STREAM_VIDEO, 0);
+    switch(CMediaSettings::GetInstance().GetCurrentVideoSettings().m_StereoMode)
+    {
+      case RENDER_STEREO_MODE_SPLIT_VERTICAL:   s.stereo_mode = "left_right"; break;
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: please use a different coding style

1) vertical alignment sooner or later breaks because there is always some variable to add that is longer than the ones before
2) only a single statement per line or setting break points is a nightmare

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 2]
```diff
@@ -3562,6 +3562,20 @@ bool CVideoPlayer::OpenVideoStream(CDVDStreamInfo& hint, bool reset)
   if (hint.stereo_mode.empty())
     hint.stereo_mode = CStereoscopicsManager::GetInstance().DetectStereoModeByString(m_item.GetPath());
 
+  {
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: why this extra block?

File: xbmc/guilib/StereoscopicsManager.cpp [hunk 3]
```diff
@@ -509,6 +509,8 @@ void CStereoscopicsManager::OnPlaybackStarted(void)
   STEREOSCOPIC_PLAYBACK_MODE playbackMode = (STEREOSCOPIC_PLAYBACK_MODE) CSettings::GetInstance().GetInt(CSettings::SETTING_VIDEOPLAYER_STEREOSCOPICPLAYBACKMODE);
   RENDER_STEREO_MODE mode = GetStereoMode();
 
+  g_infoManager.UpdateAVInfo();
```
Comment (FernetMenta) [inline xbmc/guilib/StereoscopicsManager.cpp]: UpdateAVInfo is called in CGUIWindowManager::FrameMove. means on every cycle of the main loop. why is this needed?

Comment (popcornmix) [inline xbmc/guilib/StereoscopicsManager.cpp]: I think because main thread gets blocked at start of file due to the database accesses...

Comment (FernetMenta) [inline xbmc/guilib/StereoscopicsManager.cpp]: I don't think so. if the render thread is blocked, gui won't be updated either.

Comment (FernetMenta) [inline xbmc/guilib/StereoscopicsManager.cpp]: this is gui info which is only for displaying things like 3D icons on the OSD. how is this related to the switching issue?

Comment (popcornmix) [inline xbmc/guilib/StereoscopicsManager.cpp]: We need g_infoManager.EvaluateBool("videoplayer.isstereoscopic") to be true for SetStereoMode() to be called. Without this PR that only sometimes gets called.

**Comment by popcornmix** [comment]
Added a formatting fixup commit.

File: xbmc/guilib/StereoscopicsManager.cpp (hunk 3 above)
Comment (FernetMenta) [inline xbmc/guilib/StereoscopicsManager.cpp]: > We need g_infoManager.EvaluateBool("videoplayer.isstereoscopic") to be true for SetStereoMode() to be called. 

please fix this. no behaviour is supposed to rely on gui info

Comment (popcornmix) [inline xbmc/guilib/StereoscopicsManager.cpp]: That may need some guidance.
g_graphicsContext.GetStereoMode() is used through application and renderer as the current 3D mode. The only place it gets set is from CStereoscopicsManager::ApplyStereoMode which comes from CStereoscopicsManager::OnPlaybackStarted though infomanager.

Can you explain how this information should flow?

Comment (da-anda) [inline xbmc/guilib/StereoscopicsManager.cpp]: if there is a way to directly ask the video player if the currently played file is 3D, then that would ofc be the better solution. Going the way via GUIinfoManager was the only way I had seen back then, and Infomanager had that info anyways for skin purposes.
Otoh, the 3D toggle logic should be implemented differently anyways, like in a way that won't in worst case trigger two resolution switches in a row. This however was (is) waaaaaay beyond my knowledge and skills and I didn't want to somewhat UI related code to the dvdplayer back then, which is why I went the way with messages.

Comment (FernetMenta) [inline xbmc/guilib/StereoscopicsManager.cpp]: > Going the way via GUIinfoManager was the only way I had seen back then

are you sure? GUIInfoManger needs to ask player too and it calls
g_application.m_pPlayer->GetVideoStreamInfo

Comment (da-anda) [inline xbmc/guilib/StereoscopicsManager.cpp]: or, alternatively to asking the videoplayer (which is probably a bad idea architecture wise), maybe fire off a new signal that either passes the fileitem or it's stream details around? Question then still is how to prevent the double resolution change (first change when playback is started, second when 3D mode is enabled)

Comment (da-anda) [inline xbmc/guilib/StereoscopicsManager.cpp]: > are you sure? GUIInfoManger needs to ask player too and it calls
g_application.m_pPlayer->GetVideoStreamInfo

ok, fair point. No idea why I chose that route then.

Comment (FernetMenta) [inline xbmc/guilib/StereoscopicsManager.cpp]: I think cutting the dependency to GUIInfoManager is good enough for this pr. Thoughts?

Comment (popcornmix) [inline xbmc/guilib/StereoscopicsManager.cpp]: So stereoscopicmanager calls g_application.m_pPlayer->GetVideoStreamInfo directly and avoids GUIInfoManger?
Can try that.

Comment (da-anda) [inline xbmc/guilib/StereoscopicsManager.cpp]: yeah, that's for sure the better aproach and ofc sufficient for this PR. No idea why I relied on GUI info for runtime decisions. Looking at it now that was really a bad idea.

However, @FernetMenta if you have a suggestion on how to handle/fix the resolution switch issue, I'm all ears. Also a new message/signal earlier in the process (like when all metadata has been collected/demuxed etc) would be good, because right now it can happen that ~1sec of the movie is already played before we ask for the prefered playback mode (2D vs 3D). Shall I open a thread in videoplayer dev section about it?

Comment (FernetMenta) [inline xbmc/guilib/StereoscopicsManager.cpp]: I am not too much experienced with stereoscopicmanager but I would try this.

Comment (FernetMenta) [inline xbmc/guilib/StereoscopicsManager.cpp]: cross post :)
@da-anda can you put in words what stereoscopicmanageris supposed to do?

Comment (da-anda) [inline xbmc/guilib/StereoscopicsManager.cpp]: well, it basically does two things. First task is to apply the users stereoscopic preferences to video playback (and once/if ever supported then also to pictures). I tried to do this as least intrusive as possible and thus used messages/signals to achieve this. It should also be the only entity that handles the GUI settings, so (at least in my vision) no other component should ever directly access stereoscopic related settings via the settings manager but instead always ask the manager. Once the GUI setting is being changed, the manager informs the required components about this change (curently IIRC only g_graphicsContext)

Second task is to act as connector for stereoscopic interaction between components. At least the connection from interfaces and GUI to core components. It also contains some utility functions for mode conversions (GUI enums to skiin/interfaces) that probably should have been moved to a utility class.

Comment (da-anda) [inline xbmc/guilib/StereoscopicsManager.cpp]: I forgot to describe what it currently does on video playback. When the manager detects that a stereoscopic video is being played (or user toggles from 2D to 3D etc), it tells the GUI to switch into 3D mode as well (if desired by users). GraphicsContext then does it's job and triggers a resolution switch if needed (at least on the PI when using full frame 3D modes)

Comment (FernetMenta) [inline xbmc/guilib/StereoscopicsManager.cpp]: the problem here is that this scenario can't be implemented with asynchronous messages. I would expects something like this:

- player is started with a file item
- player detects 3D
- player sets appropriate mode -> give this task to steroscopicmgr
- player waits until mode is set
- playback starts

Comment (da-anda) [inline xbmc/guilib/StereoscopicsManager.cpp]: how would player be informed when to proceed and by whom? Manager only tells graphicsContext to apply a 3D mode, it does not know when the mode actually got applied/resolution switched. Also, will graphicsContext at this stage already know about video playback metadata in order to select the correct refresh rate f.e. when switching from 3D to 2D, or who is responsible to inform it about the correct resolution and refresh rate? (I'm not into that code)

Comment (FernetMenta) [inline xbmc/guilib/StereoscopicsManager.cpp]: > how would player be informed when to proceed and by whom?

as mentioned, messages need to be synchronous. player instructs stereoscopicmgr to set mode and provides necessary information including refresh rate. s-mgr does whatever needs to be done and replies to the message. now player starts playback

**Comment by popcornmix** [comment]
Updated to not use infomanager. I'll have a look into avoiding the second hdmi mode change.

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 4]
```diff
@@ -3562,6 +3562,24 @@ bool CVideoPlayer::OpenVideoStream(CDVDStreamInfo& hint, bool reset)
   if (hint.stereo_mode.empty())
     hint.stereo_mode = CStereoscopicsManager::GetInstance().DetectStereoModeByString(m_item.GetPath());
 
+  SelectionStream& s = m_SelectionStreams.Get(STREAM_VIDEO, 0);
```
Comment (da-anda) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: shouldn't the hint also be updated according to videoSettings? And this section looks somewhat familiar, but I don't recall where it has been used. Only mentioned this so that the logic could probably been moved to some utility class if used more than once. OTOH it's not a really complex logic.

Comment (popcornmix) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: The familiar code is here: https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/VideoPlayerVideo.cpp#L779
I believe I couldn't call into that at this point in time, but I could have another try.

**Comment by popcornmix** [comment]
I've added three commits to avoid the double hdmi mode changes.
The commits aim to make the existing hdmi mode changes also switch 3d.
Switching 3D mode through the OSD will now cause a TriggerUpdateResolution.

I'm happy to hear about better ways of achieving this.

File: xbmc/guilib/GraphicContext.cpp [hunk 5]
```diff
@@ -983,16 +984,17 @@ void CGraphicContext::SetMediaDir(const std::string &strMediaDir)
   m_strMediaDir = strMediaDir;
 }
 
+
+void CGraphicContext::SetStereoMode(RENDER_STEREO_MODE mode)
```
Comment (FernetMenta) [inline xbmc/guilib/GraphicContext.cpp]: I don't see this used anywhere

File: xbmc/guilib/GraphicContext.cpp [hunk 6]
```diff
@@ -369,19 +369,19 @@ bool CGraphicContext::IsValidResolution(RESOLUTION res)
 }
 
 // call SetVideoResolutionInternal and ensure its done from mainthread
-void CGraphicContext::SetVideoResolution(RESOLUTION res, bool forceUpdate)
+void CGraphicContext::SetVideoResolution(RESOLUTION res, bool forceUpdate /*= false*/, bool switch3d /*= false*/)
```
Comment (FernetMenta) [inline xbmc/guilib/GraphicContext.cpp]: why the flag? doesn't the mode itself include whether it is a 3d mode or not?

File: xbmc/guilib/GraphicContext.cpp [hunk 7]
```diff
@@ -983,16 +984,17 @@ void CGraphicContext::SetMediaDir(const std::string &strMediaDir)
   m_strMediaDir = strMediaDir;
 }
 
+
+void CGraphicContext::SetStereoMode(RENDER_STEREO_MODE mode)
+{
+  m_stereoMode = mode;
+  g_application.m_pPlayer->TriggerUpdateResolution();
```
Comment (FernetMenta) [inline xbmc/guilib/GraphicContext.cpp]: gfx context must not have a dependency to player

**Comment by da-anda** [comment]
should we split the res-switch fix to not block the initial PR and get it merged and probably backported to Jarvis?

File: xbmc/guilib/GraphicContext.cpp (hunk 6 above)
Comment (popcornmix) [inline xbmc/guilib/GraphicContext.cpp]: No. I did originally implement separate 3D resolutions but it didn't work well. You get a huge list of extra resolutions in system/video output. Each 3D format has extra calibration data, so calibrating subtitle height for 1080p24 needs to be done again for 1080p24 SBS and 1080p24 TAB and 1080p25 FP.
In reality the TAB/SBS flags are just out-of-band AV Infoframe signalling. It is just requested as a modifier (https://github.com/xbmc/xbmc/blob/master/xbmc/windowing/egl/EGLNativeTypeRaspberryPI.cpp#L254) to the hdmi mode.

The reason for this flag is we get an unwanted resolution change from the fullscreen message which occurs after playback starts, and so after we believe we are in 3D mode, but before the renderer configure which knows the correct framerate. That means HDMI does:
1080p60 2D -> 1080p60 3D -> 1080p24 3D
and I want a way to ignore the 1080p60 3D.

**Comment by popcornmix** [comment]
@da-anda Okay - split the mode switching commit into new PR: https://github.com/xbmc/xbmc/pull/8774

**Comment by da-anda** [comment]
thanks for splitting. @FernetMenta, is it good to go from your end? I have no strong feelings about the small code dupe.

**Comment by FernetMenta** [comment]
if CStereoscopicsManager::GetVideoStereoMode is not called too often like in the main loop it is fine

**Comment by popcornmix** [comment]
A log message in CStereoscopicsManager::GetVideoStereoMode only triggered twice when playing then stopping a 3d video.

I've added a commit to avoid code duplication - is that preferable @FernetMenta ?

**Comment by FernetMenta** [comment]
looks good to me

**Comment by popcornmix** [comment]
squashed. jenkins build this please
